src/misc/unicode.cpp

```diff
@@ -451,712 +653,878 @@ static bool wide_to_code_page(const std::vector<uint16_t> &str_in,
+	error_parsing(file_name, line_num, "code page already defined");
 }
 
-static bool check_import_status(const std::ifstream &in_file,
-                                const std::string &file_name, const bool empty)
+static void error_code_page_none(const std::string& file_name, const size_t line_num)
 {
-    if (in_file.fail() && !in_file.eof()) {
-        LOG_ERR("UTF8: Error reading mapping file %s", file_name.c_str());
-        return false;
-    }
-
-    if (empty) {
-        LOG_ERR("UTF8: Mapping file %s has no entries", file_name.c_str());
-        return false;
-    }
+	error_parsing(file_name, line_num, "not currently defining a code page");
+}
 
-    return true;
+static bool check_import_status(const std::ifstream& in_file,
+                                const std::string& file_name, const bool empty)
+{
+	if (in_file.fail() && !in_file.eof()) {
+		LOG_ERR("UNICODE: Error reading mapping file %s", file_name.c_str());
+		return false;
+	}
+
+	if (empty) {
+		LOG_ERR("UNICODE: Mapping file %s has no entries",
+		        file_name.c_str());
+		return false;
+	}
+
+	return true;
 }
 
-static bool check_grapheme_valid(const Grapheme &grapheme,
-                                 const std::string &file_name, const size_t line_num)
+static bool check_grapheme_valid(const Grapheme& grapheme,
+                                 const std::string& file_name, const size_t line_num)
 {
-    if (grapheme.IsValid())
-        return true;
+	if (grapheme.IsValid()) {
+		return true;
+	}
+
+	LOG_ERR("UNICODE: Error, invalid grapheme defined in file %s, line %d",
+	        file_name.c_str(),
+	        static_cast<int>(line_num));
+	return false;
+}
 
-    LOG_ERR("UTF8: Error, invalid grapheme defined in file %s, line %d",
-            file_name.c_str(), static_cast<int>(line_num));
-    return false;
+static bool import_mapping_code_page(const std_fs::path& path_root,
+                                     const std::string& file_name,
+                                     code_page_mapping_reverse_t& mapping)
+{
+	// Import code page character -> UTF-8 mapping from external file
+
+	// Open the file
+	auto in_file = open_mapping_file(path_root, file_name);
+	if (!in_file) {
+		LOG_ERR("UNICODE: Error opening mapping file %s", file_name.c_str());
+		return false;
+	}
+
+	// Read and parse
+	std::string line_str = " ";
+	size_t line_num      = 0;
+
+	code_page_mapping_reverse_t new_mapping;
+
+	while (get_line(in_file, line_str, line_num)) {
+		std::vector<std::string> tokens;
+		if (!get_tokens(line_str, tokens)) {
+			continue; // empty line
+		}
+
+		uint8_t character_code = 0;
+		if (!get_hex_8bit(tokens[0], character_code)) {
+			error_parsing(file_name, line_num);
+			return false;
+		}
+
+		if (GCC_UNLIKELY(tokens.size() == 1)) {
+			// Handle undefined character entry, ignore 7-bit ASCII
+			// codes
+			if (character_code >= decode_threshold_non_ascii) {
+				Grapheme grapheme;
+				add_if_not_mapped(new_mapping, character_code, grapheme);
```

Comment:
I wasn't sure how this was being used because it goes out of scope.
Suggest: `const Grapheme grapheme; // placeholder`

Reply:
Ok, changed.